Docs: Updates the release and maintenance page to reflect new cadences (#73781)
ansible/ansible: docs/docsite/rst/reference_appendices/release_and_maintenance.rst

.. _release_and_maintenance:

************************
Releases and maintenance
************************

This section describes release cycles, rules, and maintenance schedules for both Ansible community projects: the Ansible community package and ``ansible-core``. The two projects have different versioning systems, maintenance structures, contents, and workflows.

> ====================================================  =================================================

Reviewer: I like this comparison table, good idea.

Ansible community package                             ansible-core
====================================================  =================================================

> each major release is a new major version (4.0.0)     continues "classic Ansible" versioning (2.11)

Reviewer: I wonder if this might make it clearer what a major version is:
```suggestion
each major release is a new major version (3.0.0 -> 4.0.0)     continues "classic Ansible" versioning (2.10 -> 2.11)
```

Author: Good idea. I'll try to implement this, but not through the suggestion here, because it will require updates to the table layout

Author: Addressed in cca19eef3206585b6f806ace09cba21a993cab28.

follows semantic versioning rules                     does not use semantic versioning
maintains only one version at a time                  maintains latest version plus two older versions

> includes language, runtime, and selected Collections  includes only the language and runtime

Reviewer 2: ```suggestion
includes language, runtime, and builtin Collections  includes only the language and runtime
```

> developed and maintained in Collection repositories   developed and maintained in ansible/ansible repo

Reviewer 2: ```suggestion
developed and maintained in Collection repositories   developed and maintained in ansible/ansible repository
```

Author: this one will also cause a test failure by running over the table width, but I will incorporate it when I adopt gundalow's suggestion above

Author: Addressed in cca19eef3206585b6f806ace09cba21a993cab28.

> developed and maintained in Collection repositories   developed and maintained in ansible/ansible repo

Reviewer 3: @acozine 
I think you should use sentence-style capitalization in the table text.

Author: thanks @lmalivert - addressed in cca19eef3206585b6f806ace09cba21a993cab28

Reviewer 3: @acozine LGTM 👍🏽 Thanks!

====================================================  =================================================

> Most community users install the Ansible community package. The ``ansible-core`` option is primarily for developers.

Reviewer 2: ```suggestion
Most community users install the Ansible community package. The ``ansible-core`` option is primarily for developers and users who want to install only collections they need.
```
Or something like that. I think to install `ansible-core` + collections that and only that needed from Galaxy is a very good approach for some users, IMO

> Most community users install the Ansible community package. The ``ansible-core`` option is primarily for developers.

Reviewer 4: Is this really the message we want to send out? I see the `ansible-core` package plus needed collections as the way forward. The "we include everything and the kitchen sink" approach is in my books a stopgap measure that gives users time to migrate their stuff.

Reviewer 2: @tadeboro what do you think of https://github.com/ansible/ansible/pull/73781#discussion_r587332303 ?

.. contents::
   :local:

.. _release_cycle:

Release cycle overview
======================

The two community releases are related - the release cycle follows this pattern:

#. Release of a new ansible-core major version, for example, ansible-core 2.11

>    * New release of ansible-core and two prior versions are now maintained (ansible-base 2.10, Ansible 2.9)

Reviewer 2: ```suggestion
   * New release of ansible-core and two prior versions are now maintained (in this case, ansible-base 2.10, Ansible 2.9)
```

   * Work on new features for ansible-core continues in the ``devel`` branch

#. Collection freeze (no new Collections or new versions of existing Collections) on the Ansible community package
#. Release candidate for Ansible community package, testing, additional release candidates as necessary
#. Release of a new Ansible community package major version based on the new ansible-core, for example, Ansible 4.0.0 based on ansible-core 2.11

   * Newest release of the Ansible community package is the only version now maintained
   * Work on new features continues in Collections

>    * Individual Collections can make multiple minor and/or major releases

Reviewer 2: ```suggestion
   * Individual Collections can make multiple minor and/or major releases
```
I think it's worth mentioning somewhere (not exactly here) that an Ansible community package major release includes latest major collection releases and keep them until next Ansible community package major release, so only minor and bugfix collection releases will be accepted between Ansible community package major releases.
@felixfontein , am i correct?

Reviewer 2: I can see a bit below something about it but it doesn't seem to be explicitly said

Reviewer 2: I changed thttps://github.com/ansible/ansible/pull/73781#discussion_r587344064 a little bit to clarify this

Reviewer 2: I changed https://github.com/ansible/ansible/pull/73781#discussion_r587344064 a little bit to clarify this

#. Minor releases of three maintained ansible-core versions every three weeks (2.11.1)

> #. Minor releases of the single maintained Ansible community package version every three weeks (4.1.0)

Reviewer 5: Are we timing Ansible package minor release to always pick up the ansible-core minor release? As in can a user assume Ansible 4.x.0 will have ansible-core 2.11.x ? (one to one correspondence?) or is it too tricky to commit to syncing up the minor releases like this?

Author: We haven't committed to updating the included version of ansible-core on a specific schedule, so I think we shouldn't mention anything specific here.

#. Feature freeze on ansible-core
#. Release candidate for ansible-core, testing, additional release candidates as necessary
#. Release of the next ansible-core major version, cycle begins again

Release cycle for the Ansible community package
-----------------------------------------------

> The Ansible community team typically releases two major versions of the community package per year, on a flexible release cycle that trails the release of ``ansible-core``. This cycle can be extended to allow for larger changes to be properly implemented and tested before a new release is made available. See :ref:`roadmaps` for upcoming release details. Between major versions we release a new minor version of the Ansible community package every three weeks. Minor releases include new backwards-compatible features, modules and plugins as well as bug fixes.

Reviewer 5: ```suggestion
The Ansible community team typically releases two major versions of the community package per year, on a flexible release cycle that trails the release of ``ansible-core``. This cycle can be extended to allow for larger changes to be properly implemented and tested before a new release is made available. See :ref:`roadmaps` for upcoming release details. Between major versions, we release a new minor version of the Ansible community package every three weeks. Minor releases include new backwards-compatible features, modules and plugins, as well as bug fixes.
```

Starting with version 2.10, the Ansible community team guarantees maintenance for only one major community package release at a time. For example, when Ansible 4.0.0 gets released, the team will stop making new 3.x releases. Community members may maintain older versions if desired.

> Older, unmaintained versions of the Ansible community package might contain unfixed security vulnerabilities (*CVEs*). If you are using a release of the Ansible community package that is no longer maintained, we strongly encourage you to upgrade as soon as possible in order to benefit from the latest features and security fixes.

Reviewer 5: Consider turning this into a note as it seems important to highlight.

Author: Good thought, done.

> Each major release of the Ansible community package accepts the latest released version of each included Collection and the latest released version of ansible-core. Major releases of the Ansible community package can contain breaking changes in the modules and other plugins within the included Collections and/or in core features.

Reviewer 5: The Collection version is fixed at RC time. Only a blocker would allow a collection version update at that point afaik. Ask @abadger for sure.

So maybe the first sentence needs something like "latest released version of each included Collection at the time the release candidate happens". Or something. Dunno - the wording addition helps developers understand when their collection will 'freeze'. Not sure if the added sentence helps users though, so take  this suggestion w/ a grain of salt.

Author: Yes, good point. I added a link to the Roadmaps "for specific schedules and deadlines" so this page can stay a little more general.

> The Ansible community package follows semantic versioning rules. Minor releases of the Ansible community package accept only backwards-compatible changes in included Collections. Collections must also use semantic versioning, so the Collection version numbers reflect this rule. For example, if Ansible 3.0.0 releases with community.general 2.0.0, then all minor releases of Ansible 3.x (such as Ansible 3.1.0 or Ansible 3.5.0) must include a 2.x release of community.general (such as 2.8.0 or 2.9.5).

Reviewer 2: ```suggestion
The Ansible community package follows semantic versioning rules. Minor releases of the Ansible community package accept only backwards-compatible changes in included Collections, in other words, not Collections major releases. Collections must also use semantic versioning, so the Collection version numbers reflect this rule. For example, if Ansible 3.0.0 releases with community.general 2.0.0, then all minor releases of Ansible 3.x (such as Ansible 3.1.0 or Ansible 3.5.0) must include a 2.x release of community.general (such as 2.8.0 or 2.9.5) and not 3.x.x or later major releases.
```
or something like that

Work in Collections is tracked within the individual Collection repositories.

> You can refer to the :ref:`porting guides<porting_guides>` for tips on updating your Ansible playbooks to run on newer versions. For Ansible 2.10 and later releases, you can install the Ansible package with ``pip``. See :ref:`intro_installation_guide` for details.  For older releases, You can download the Ansible release from `<https://releases.ansible.com/ansible/>`_.

Reviewer 2: ```suggestion
You can refer to the :ref:`porting guides<porting_guides>` for tips on updating your Ansible playbooks to run on newer versions. For Ansible 2.10 and later releases, you can install the Ansible package with ``pip``. See :ref:`intro_installation_guide` for details.  For older releases, you can download the Ansible release from `<https://releases.ansible.com/ansible/>`_.
```

This table links to the release notes for each major Ansible release. These release notes (changelogs) contain the dates and significant changes in each minor release.

==================================      =================================================
Ansible Community Package Release       Status
==================================      =================================================
`3.0 Release Notes`_                    Current
`2.10 Release Notes`_                   Unmaintained (end of life)
==================================      =================================================

.. _3.0 Release Notes:
.. _2.10 Release Notes:
.. _2.10: https://github.com/ansible-community/ansible-build-data/blob/main/2.10/CHANGELOG-v2.10.rst

> Release cycle for ansible-base/ansible-core
> -------------------------------------------

Reviewer: Do you think this might be clearer:
```suggestion
Release cycle for ansible-base & ansible-core
---------------------------------------------
```

Reviewer 2: maybe to remove `ansible-base` at all? There is only the `ansible-core` notion used in the following sentences

Author: Addressed in cca19eef3206585b6f806ace09cba21a993cab28.

``ansible-core`` is developed and released on a flexible release cycle. This cycle can be extended in order to allow for larger changes to be properly implemented and tested before a new release is made available. See :ref:`roadmaps` for upcoming release details.

``ansible-core`` has a graduated maintenance structure that extends to three major releases.
For more information, read about the :ref:`development_and_stable_version_maintenance_workflow` or
see the chart in :ref:`release_schedule` for the degrees to which current releases are maintained.

> Older, unmaintained versions of ``ansible-core`` can contain unfixed security vulnerabilities (*CVEs*). If you are using a release of ``ansible-core`` that is no longer maintained, we strongly encourage you to upgrade as soon as possible to benefit from the latest features and security fixes.

Reviewer 5: same suggestion on  turning this into a note to highlight its importance.

Author: Done.

> You can refer to the :ref:`porting guides<porting_guides>` for tips on updating your Ansible

Reviewer 5: fyi - this anchor goes only to Ansible porting guides now, not ansible-core.  The Ansible porting guide does include the core info so probably okay, but if the user is reading this page on the ansible-core docs, this link will hop them over to the Ansible docs (..once we update intersphinx links).

Author: I put in a link for each (one for core, one for the package, each labelled as such).

In between releases, the Ansible community develops new features, maintains existing functionality, and fixes bugs in ``ansible-core`` and in the collections included in the Ansible community package.

Workflow for the Ansible community package
------------------------------------------

The Ansible community develops and maintains the features and functionality included in the Ansible community package in Collections repositories, with a workflow that looks like this:

 * Developers add new features and bug fixes to the individual Collections, following each Collection's rules on contributing.
 * Each new feature and each bug fix includes a changelog fragment describing the work.
 * Release engineers create a minor release for the current version every three weeks to ensure that the latest bug fixes are available to users.
 * At the end of the development period, the release engineers announce which Collections, and which major version of each included Collection,  will be included in the next release of the Ansible community package. New Collections and new major versions may not be added after this, and the work of creating a new release begins.

> We generally do not provide fixes for unmaintained releases of the Ansible community package, however, there can sometimes be exceptions for critical issues.

Reviewer 3: Suggestion: `However, there are times when exceptions are made for critical issues.`

============================================= ======================================================
devel                                         In development (ansible-core 2.11 unreleased, trunk)
`2.10 ansible-base Release Notes`_            Maintained (security **and** general bug fixes)

> `2.9 Release Notes`_                          Maintained (pre-collections)(security **and** critical bug fixes)

Reviewer: ```suggestion
`2.9 Release Notes`_                          Maintained (pre-collections) (security **and** critical bug fixes)
```

`2.6 Release Notes`_                          Unmaintained (end of life)
`2.5 Release Notes`_                          Unmaintained (end of life)
<2.5                                          Unmaintained (end of life)
============================================= ======================================================

.. _2.10 ansible-base Release Notes:
.. _2.10-base: https://github.com/ansible/ansible/blob/stable-2.10/changelogs/CHANGELOG-v2.10.rst
.. _2.9 Release Notes:
.. _2.9: https://github.com/ansible/ansible/blob/stable-2.9/changelogs/CHANGELOG-v2.9.rst
.. _2.8 Release Notes:
.. _2.8: https://github.com/ansible/ansible/blob/stable-2.8/changelogs/CHANGELOG-v2.8.rst
.. _2.7 Release Notes: https://github.com/ansible/ansible/blob/stable-2.7/changelogs/CHANGELOG-v2.7.rst
.. _2.6 Release Notes:
.. _2.6: https://github.com/ansible/ansible/blob/stable-2.6/changelogs/CHANGELOG-v2.6.rst
.. _2.5 Release Notes: https://github.com/ansible/ansible/blob/stable-2.5/changelogs/CHANGELOG-v2.5.rst
.. _support_life:
.. _methods:

Preparing for a new release
---------------------------

.. _release_freezing:

Feature freezes
^^^^^^^^^^^^^^^

> While there is a pending release candidate, core developers and maintainers focus on improving the release candidate, not on adding or reviewing new features.

Reviewer 5: Ansible roadmap says feature freeze happens at beta. ansible-core roadmap does not specify a feature freeze time.

Author: I made this more general - "in the final preparations for a release" so the section is more of a general warning that we may not merge your PR immediately.

A feature freeze means that we delay new features and fixes that are not related to the release candidate so we can ship the new release as soon as possible.

Release candidates
^^^^^^^^^^^^^^^^^^^

We create at least one release candidate before each new major release of Ansible or ``ansible-core``. Release candidates allow the Ansible community to try out new features, test existing playbooks on the release candidate, and report bugs or issues they find.

Ansible and ``ansible-core`` tag the first release candidate (RC1) which is usually scheduled to last five business days. If no major bugs or issues are identified during this period, the release candidate becomes the final release.

If there are major problems with the first candidate, the team and the community fix them and tag a second release candidate (RC2). This second candidate lasts for a shorter duration than the first. If no problems have been reported for an RC2 after two business days, the second release candidate becomes the final release.

> If there are major problems in RC2, the cycle begins again with another release candidate and repeats until the core maintainers agree that all major problems have been fixed.

Reviewer 5: ```suggestion
If there are major problems in RC2, the cycle begins again with another release candidate and repeats until the maintainers agree that all major problems have been fixed.
```

Reviewer 5: core suggests ansible-core  to me... so suggesting we snip it out.

.. _development_and_stable_version_maintenance_workflow:

Development and stable version maintenance workflow
----------------------------------------------------

In between releases, the Ansible community develops and maintains ``ansible-core`` on GitHub_, with a workflow that looks like this:

 * Developers add new features and bug fixes to the ``devel`` branch.

>  * Each new feature and each notable bug fix includes a changelog fragment describing the work.

Reviewer 2: ```suggestion
 * Each new feature and each bug fix includes a changelog fragment describing the work.
```
IMO even a code refactoring should include a fragment. It can help later in case of bugs introduced

 * The development team backports bug fixes to one, two, or three stable branches, depending on the severity of the bug. They do not backport new features.
 * Release engineers create a minor release for each maintained version every three weeks to ensure that the latest bug fixes are available to users.
 * At the end of the development period, the release engineers impose a feature freeze and the work of creating a new release begins.

> We generally do not provide fixes for unmaintained releases of Ansible, however, there can sometimes be exceptions for critical issues.

Reviewer 5: Is it worth adding how the Ansible project develops, since the prior list seems focused on ansible-core? Maybe something as simple as 'The Ansible community package is developed in multiple repositories that source each collection. As such, there is no single  ``devel`` branch.

Then we can elaborate on that once we have a devel docsite for Ansible to say what it's reflecting.

Author: Yep, good idea. I split this section up so we have Workflow for core and Workflow for the package.

.. _GitHub: https://github.com/ansible/ansible

.. _release_changelogs:

Changelogs
^^^^^^^^^^

> We generate changelogs based on fragments. Here is the generated changelog for 2.9_ as an example. When creating new features or fixing bugs, create a changelog fragment describing the change. A changelog entry is not needed for new modules or plugins. Details for those items will be generated from the module documentation.

Reviewer 2: ```suggestion
We generate changelogs based on fragments. When creating new features for existing modules and plugins or fixing bugs, create a changelog fragment describing the change. A changelog entry is not needed for new modules or plugins. Details for those items will be generated from the module documentation.
```
1) IMO to put this reference https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment is enough. There are good examples
2) fragments are not needed for new modules / plugins.

Author: I think a link to a changelog is useful, but we don't need an extra sentence for it - I'll add a link to the word `changelogs` in the first sentence, and folks who are curious can use that.

> For guidance on creating changelog fragments, see the :ref:`changelog examples and instructions<changelogs_how_to>` in the Community Guide.

Reviewer 5: I don't think all collections in Ansible use our changelog generator.

Author: I added a link to the antsibull-changelog tool with a note that we recommend it. We can add more detail in that section later.

> Deprecation Cycles

Reviewer 5: ```suggestion
Deprecation cycles
```

Ansible deprecation cycle
^^^^^^^^^^^^^^^^^^^^^^^^^

Since Ansible is a package of individual collections, the deprecation cycle depends on the collection maintainers. We recommend the collection maintainers deprecate a feature in one Ansible major version and do not remove that feature for one year, or at least until the next major Ansible version. For example, deprecate the feature in 3.1.0, and do not remove the feature until 5.0.0, or 4.0.0 at the earliest. Collections should use semantic versioning, such that the major collection version cannot be changed within an Ansible major version. Thus the removal should not happen before the next major Ansible community package release. This is up to each collection maintainer and cannot be guaranteed.

ansible-core deprecation cycle
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The cycle is normally across 4 feature releases (2.x.y, where the x marks a feature release and the y a bugfix release), so the feature is normally removed in the 4th release after we announce the deprecation. For example, something deprecated in 2.9 will be removed in 2.13, assuming we don't jump to 3.x before that point. The tracking is tied to the number of releases, not the release numbering.

Reviewer 2: ```suggestion
The cycle is normally across 4 feature releases (2.x.y, where the x marks a feature release and the y a bugfix release), so the feature is normally removed in the 4th release after we announce the deprecation. For example, something deprecated in 2.9 will be removed in 2.13, assuming we do not jump to 3.x before that point. The tracking is tied to the number of releases, not the release numbering.
```
just to follow the "best practices" guide

Author: ah, thanks, I missed that one

> For modules/plugins that remain in ansible-core, we keep the documentation after the removal for users of older versions.

Reviewer 5: do we? I mean we can say in general we do even for Ansible because 2.10 docs aren't going anywhere.. .so not sure what this sentence is telling me?

Author: Good catch, I'm not sure what this is meant to convey. It can probably just get removed.

Author: Removed.
